kupyna/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+// #![no_std]
+#![doc = include_str!("../README.md")]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/RustCrypto/media/6ee8e381/logo.svg",
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/media/6ee8e381/logo.svg"
+)]
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+#![forbid(unsafe_code)]
+#![warn(missing_docs)]
+
+pub use digest::{self, Digest};
+
+use core::{convert::TryInto, fmt};
+#[cfg(feature = "zeroize")]
+use digest::zeroize::{Zeroize, ZeroizeOnDrop};
+use digest::{
+    HashMarker, InvalidOutputSize, Output,
+    block_buffer::Eager,
+    core_api::{
+        AlgorithmName, Block, BlockSizeUser, Buffer, BufferKindUser, CoreWrapper,
+        CtVariableCoreWrapper, OutputSizeUser, RtVariableCoreWrapper, TruncSide, UpdateCore,
+        VariableOutputCore,
+    },
+    crypto_common::hazmat::{DeserializeStateError, SerializableState, SerializedState},
+    typenum::{U6, U28, U32, U48, U64, U72, U128, U136, Unsigned},
+};
+
+mod compress1024;
+mod compress512;
+mod tables;
+
+/// Lowest-level core hasher state of the short Kupyna variant.
+#[derive(Clone)]
+pub struct KupynaShortVarCore {
+    state: [u64; compress512::COLS],
+    blocks_len: u64,
+}
+
+/// Short Kupyna variant which allows to choose output size at runtime.
+pub type KupynaShortVar = RtVariableCoreWrapper<KupynaShortVarCore>;
+/// Core hasher state of the short Kupyna variant generic over output size.
+pub type KupynaShortCore<OutSize> = CtVariableCoreWrapper<KupynaShortVarCore, OutSize>;
+/// Hasher state of the short Kupyna variant generic over output size.
+pub type KupynaShort<OutSize> = CoreWrapper<KupynaShortCore<OutSize>>;
+
+/// Kupyna-48 hasher state
+pub type Kupyna48 = CoreWrapper<KupynaShortCore<U6>>;
+/// Kupyna-224 hasher state.
+pub type Kupyna224 = CoreWrapper<KupynaShortCore<U28>>;
+/// Kupyna-256 hasher state.
+pub type Kupyna256 = CoreWrapper<KupynaShortCore<U32>>;
+
+impl HashMarker for KupynaShortVarCore {}
+
+impl BlockSizeUser for KupynaShortVarCore {
+    type BlockSize = U64;
+}
+
+impl BufferKindUser for KupynaShortVarCore {
+    type BufferKind = Eager;
+}
+
+impl UpdateCore for KupynaShortVarCore {
+    #[inline]
+    fn update_blocks(&mut self, blocks: &[Block<Self>]) {
+        self.blocks_len += blocks.len() as u64;
+        for block in blocks {
+            compress512::compress(&mut self.state, block.as_ref());
+        }
+    }
+}
+
+impl OutputSizeUser for KupynaShortVarCore {
+    type OutputSize = U32;
+}
+
+impl VariableOutputCore for KupynaShortVarCore {
+    const TRUNC_SIDE: TruncSide = TruncSide::Right;
+
+    #[inline]
+    fn new(output_size: usize) -> Result<Self, InvalidOutputSize> {
```

Comment:
So it looks like `output_size` does not change generated hash value outside of simple truncation? Frankly, it's quite unusual... But we have no choice but to follow the spec here.